🐛 Fix nodes outbound connectivity with SLB

CecileRobertMichon opened the pull request:
**What this PR does / why we need it**: This implements outbound connectivity for all nodes as described in https://github.com/kubernetes-sigs/cloud-provider-azure/tree/master/docs/services#standard-loadbalancer. For each Azure cluster, a new node outbound public IP and public Load Balancer will be created. The Load Balancer's name is the cluster name so that cloud provider can detect the LB exists and reuse the same LB for exposing services. This LB will be created with a backend pool and each node machine (all VMs that are not control-planes) will be added to this backend pool on its creation.  The LB will also have outbound rules defined to allow UDP and TCP outbound traffic from all nodes, as described in[ Azure docs](https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-outbound-connections#proto).

See also:
https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-outbound-connections

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially addresses #648 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Fix nodes outbound connectivity with Standard Load Balancer
```

CecileRobertMichon commented on cloud/services/publicloadbalancers/publicloadbalancers.go:
```diff
@@ -33,6 +33,7 @@ import (
 type Spec struct {
 	Name         string
 	PublicIPName string
+	Role         string
```
@jsturtevant please review if you can as these changes will affect #646 (positively, I hope)

jsturtevant replied:
Looks like a nice improvement 👍

jsturtevant commented on cloud/services/publicloadbalancers/publicloadbalancers.go:
```diff
@@ -72,89 +76,94 @@ func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
+					Name: to.StringPtr("OutboundNATAllProtocols"),
+					OutboundRulePropertiesFormat: &network.OutboundRulePropertiesFormat{
+						Protocol:             network.LoadBalancerOutboundRuleProtocolAll,
+						IdleTimeoutInMinutes: to.Int32Ptr(4),
```
In aks engine this is set to 30 mins: 

https://github.com/Azure/aks-engine/blob/95d6127337008ed562d3723a53dbffbc99b2eee2/pkg/api/defaults.go#L421
https://github.com/Azure/aks-engine/blob/108dfabdaaf049b3062d9b6298e4ca03b9b3473e/pkg/api/const.go#L229

Maybe @jackfrancis can confirm?

jackfrancis replied:
Correct, and it's also user-configurable. I can't speak to whether 4 or 30 is a preferable default. I do wonder if it should be user-configurable.

What does this outbound rule timeout govern?

CecileRobertMichon replied:
4 is the default value https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-outbound-connections#idletimeout and is also the same value CAPZ is already using so didn't want to change the default on this PR (this block of code just moved down).
We can look at making it configurable if there is a use case for configuring that value separately.

jsturtevant replied:
The comment above the linked line seems like it might have been a bug at some point?

>// DefaultOutboundRuleIdleTimeoutInMinutes determines the aks-engine provided default for IdleTimeoutInMinutes of the OutboundRule of the agent loadbalancer
	// This value is set greater than the default Linux idle timeout (15.4 min): https://pracucci.com/linux-tcp-rto-min-max-and-tcp-retries2.html

CecileRobertMichon replied:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/676

jackfrancis replied:
Understood. I'm not sure why the default is set so high in AKS Engine, as it would seem to have the side-effect of reducing port availability (leading to exhaustion).

In any event, having read about what this does I think the proper value of this is going to be very much app domain-specific, so: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/677

devigned commented on cloud/defaults.go:
```diff
@@ -79,6 +79,11 @@ func GeneratePublicIPName(clusterName, hash string) string {
 
+// GenerateNodeOutboundIPName generates a public IP name, based on the cluster name.
+func GenerateNodeOutboundIPName(clusterName string) string {
+	return fmt.Sprintf("%s-node-ip-outbound", clusterName)
```
We may want to think about resource naming conventions: https://docs.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/naming-and-tagging.

See also https://kubernetes.slack.com/archives/CEX9HENG7/p1590166717280400

CecileRobertMichon replied:
Good point, thanks for pointing me to this resource (didn't know it existed). I will fix this name, and I opened https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/678 for broader auditing of resource naming.

CecileRobertMichon replied:
renamed to follow best practices but I kept the cluster name in there for now because it technically allows having multiple CAPZ clusters in one RG without conflict since public IPs are scoped to the RG. In the future, we should determine whether that's a real requirement and, if not, clearly document that each cluster has to be in its own RG (so we can remove the cluster-name / resource naming dependency).

jsturtevant commented on cloud/services/publicloadbalancers/publicloadbalancers.go:
```diff
@@ -72,89 +76,94 @@ func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
+		}
+	}
+
+	err = s.Client.CreateOrUpdate(ctx,
```
nit: not sure what the formatting rules are but this feels like it should be on a single line

CecileRobertMichon replied:
fixed, this was leftover from when lb was the huge struct definition

CecileRobertMichon commented:
/hold

I found a potential bug, investigating...

CecileRobertMichon commented:
ok so the bug is that cloud-provider can't find the VMSS instances to add them to the backend pool because host name != instance name, not related to this PR (it's already in master). I will work on a fix separately, this can go in as-is for now since it fixes the other issue.

/hold cancel

CecileRobertMichon commented:
The apidiff failure is due to:

```
sigs.k8s.io/cluster-api-provider-azure/cloud/services/scalesets
  Compatible changes:
  - Service.PublicLoadBalancersClient: added
  - Spec.PublicLoadBalancerName: added
sigs.k8s.io/cluster-api-provider-azure/cloud/services/networkinterfaces
  Compatible changes:
  - Spec.MachineRole: added
sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3
  Incompatible changes:
  - APIServerRoleTagValue: removed
  - BastionRoleTagValue: removed
  - CommonRoleTagValue: removed
  - PrivateRoleTagValue: removed
  - PublicRoleTagValue: removed
  Compatible changes:
  - APIServerRole: added
  - BastionRole: added
  - CommonRole: added
  - NodeOutboundRole: added
  - PrivateRole: added
  - PublicRole: added
sigs.k8s.io/cluster-api-provider-azure/cloud/services/publicloadbalancers
  Compatible changes:
  - Spec.Role: added
sigs.k8s.io/cluster-api-provider-azure/cloud
  Compatible changes:
  - GenerateNodeOutboundIPName: added
```

No API breaking changes.

CecileRobertMichon commented:
fixed the other issue in #680
